pkg/ring/ring.go

```diff
@@ -363,29 +363,73 @@ func (r *Ring) GetReplicationSetForOperation(op Operation) (ReplicationSet, erro
 		return ReplicationSet{}, ErrEmptyRing
 	}
 
-	// Calculate the number of required ingesters;
-	// ensure we always require at least RF-1 when RF=3.
-	numRequired := len(r.ringDesc.Ingesters)
-	if numRequired < r.cfg.ReplicationFactor {
-		numRequired = r.cfg.ReplicationFactor
-	}
-	maxUnavailable := r.cfg.ReplicationFactor / 2
-	numRequired -= maxUnavailable
-
-	ingesters := make([]IngesterDesc, 0, len(r.ringDesc.Ingesters))
+	// Build the initial replication set, excluding unhealthy instances.
+	instances := make([]IngesterDesc, 0, len(r.ringDesc.Ingesters))
+	zoneFailures := make(map[string]struct{})
 	for _, ingester := range r.ringDesc.Ingesters {
 		if r.IsHealthy(&ingester, op) {
-			ingesters = append(ingesters, ingester)
+			instances = append(instances, ingester)
+		} else {
+			zoneFailures[ingester.Zone] = struct{}{}
 		}
 	}
 
-	if len(ingesters) < numRequired {
-		return ReplicationSet{}, ErrTooManyFailedIngesters
+	// Max errors and max unavailable zones are mutually exclusive. We initialise both
+	// to 0 and then we update them whether zone-awareness is enabled or not.
+	maxErrors := 0
+	maxUnavailableZones := 0
+
+	if r.cfg.ZoneAwarenessEnabled {
+		// Given data is replicated to RF different zones, we can tolerate a number of
+		// RF/2 failing zones. However, we need to protect from the case the ring currently
+		// contains instances in a number of zones < RF.
+		numReplicatedZones := util.Min(len(r.ringZones), r.cfg.ReplicationFactor)
+		minSuccessZones := (numReplicatedZones / 2) + 1
+		maxUnavailableZones = numReplicatedZones - minSuccessZones
+
+		if len(zoneFailures) > maxUnavailableZones {
+			return ReplicationSet{}, ErrTooManyFailedIngesters
+		}
+
+		if len(zoneFailures) > 0 {
+			// We remove all instances (even healthy ones) from zones with at least
+			// 1 failing ingester. Due to how replication works when zone-awareness is
+			// enabled (data is replicated to RF different zones), there's no benefit in
+			// querying healthy instances from "failing zones".
+			filteredInstances := make([]IngesterDesc, 0, len(r.ringDesc.Ingesters))
```

Comment:
The tokens in a zone are randomly distributed. So if one of the ingesters in a zone fails the only way to make sure you have all the data is to loop over all the tokens and see if there is data for all of them. We're trying to avoid having to do that computation.
If all the ingesters in a zone are healthy then we don't have to worry about that.

Reply:
I commented on this to improve the comment in the code.

Comment in the code says "there's no benefit querying healthy instances from failing zones", but your explanation suggests that it would actually be harmful (too slow). I don't quite see why we would need to loop over all tokens – we know which ingester(s) "own" the token (main + replicas), and we only need to ask those.

I think the real reason why we can skip all ingesters in this zone, is that we either treat entire zone as failed if there is single error, or successul if there are no errors.

(Emphasis on "I think". I may be wrong about it)

Reply:
The comment was updated.

Reply:
> I don't quite see why we would need to loop over all tokens – we know which ingester(s) "own" the token (main + replicas), and we only need to ask those.

I also don't understand why we would need to loop over all of them.